### PR TITLE
Helm chart

### PR DESCRIPTION
**Pull request by bikash119**
# Description

- Use helm chart to manage k8s resources
- Added pre-commit hook to use `helm lint` to find and fix linting errors in chart templates.

Closes #<issue_number>

**Type of change**

- Improvement (change adding some improvement to an existing functionality)
- Documentation update

**How Has This Been Tested**

on dev machine
- install minikube
- install helm
- helm install <name> <chart_folder_location> . In this case `helm install argilla-dev examples/deployment/k8s/argilla-cart`
This should install and start elasticsearch and argilla-server to the k8s cluster
- `kubectl port-forward service/argilla-dev-argilla-chart 6900:6900`. 
- http://localhost:6900 ->should open the Argilla UI login screen.
- USERNAME: argilla , PASSWORD: 12345678, will authenticate the user and take user to http://localhost:6900/datasets.

**Checklist**

- I added relevant documentation [TODO]
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation [TODO]
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)[ TODO ]

### CODE REVIEW

**Comment by bikash119** [comment]
@davidberenstein1957 : Should we remove the k8s resources files inside examples/deployment/k8s folder once we have helm charts for Elasticsearch and Argilla? We will have hard time keeping the k8s resource yamls and chart templates in sync if we decide to keep them both. Please share your feedback.
cc: @gabrielmbmb

**Comment by davidberenstein1957** [comment]
@frascuchon

**Comment by bikash119** [comment]
hi @davidberenstein1957 , @frascuchon. Request you to share your feedback on this work as I continue to work on creating readme for this change. I have been able to resolve all the review comments for the original PR [4065](https://github.com/argilla-io/argilla/pull/4065).
In a nutshell:
- Argilla can be installed using helm chart
- Redis is added as dependency chart.
- ES is installed using Elastic Operator
- ES installation is optional based on value in values.yaml.
- All of the services use PV to persist data.
- Default Owner user can be set from values.yaml.
- modified .pre-commit-config.yml to use helm lint for k8s resource yamls.

File: examples/deployments/k8s/argilla-chart/README.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+   helm init
+   ```
+
+For more detailed installation instructions, please refer to the [official Helm documentation](https://helm.sh/docs/intro/install/).
+
+## Dependencies
+
+- Redis
+- Elasticsearch
+## Adding the Helm Repository
```
Comment (frascuchon) [inline examples/deployments/k8s/argilla-chart/README.md]:
```suggestion
- Elasticsearch

## Adding the Helm Repository
```

File: argilla-server/CHANGELOG.md [hunk 2]
```diff
@@ -15,7 +15,8 @@ These are the section headers that we use:
 -->
 
 ## [Unreleased]()
-
+### Added
+- Added helm chart for argilla. ([#5512](https://github.com/argilla-io/argilla/pull/5512))
```
Comment (frascuchon) [inline argilla-server/CHANGELOG.md]:
```suggestion

### Added

- Added helm chart for argilla. ([#5512](https://github.com/argilla-io/argilla/pull/5512))

```

File: examples/deployments/k8s/argilla-chart/README.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+```
+
+## Upgrading
+
+To upgrade the chart to a newer version:
+
+```bash
+helm repo update
+helm upgrade my-release argilla/argilla
+```
```
Comment (frascuchon) [inline examples/deployments/k8s/argilla-chart/README.md]:
```suggestion
After adding the repository, you can install the chart with the release name `my-argilla-server`:

```bash
helm install my-argilla-server examples/deployments/k8s/argilla-chart
```

## Configuration

To override default values, override the values in `values.yaml` file and install the chart using:

```bash
helm install my-argilla-server examples/deployments/k8s/argilla-chart
```

## Upgrading

To upgrade the chart to a newer version:

```bash
helm repo update
helm upgrade my-argilla-server argilla/argilla
```
```

File: examples/deployments/k8s/argilla-chart/README.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+TODO: Add information about any backwards incompatible changes and how to handle them.
+
+## Uninstalling the Chart
+
+To uninstall/delete the `my-release` deployment:
```
Comment (frascuchon) [inline examples/deployments/k8s/argilla-chart/README.md]:
```suggestion
To uninstall/delete the `my-argilla-server` deployment:
```

File: examples/deployments/k8s/argilla-chart/README.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+```bash
+helm delete my-release
```
Comment (frascuchon) [inline examples/deployments/k8s/argilla-chart/README.md]:
```suggestion
helm delete my-argilla-server
```

File: examples/deployments/k8s/argilla-chart/templates/NOTES.txt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+Using external Elasticsearch at:
+  http://{{ .Values.externalElasticsearch.host }}:{{ .Values.externalElasticsearch.port }}
+{{- end }}
+
+{{- if .Values.argilla.persistence.enabled }}
+Persistence for Argilla is enabled. Your Argilla data will be stored in a Persistent Volume Claim named:
+  {{ .Release.Name }}-argilla-pvc
+{{- end }}
+
+{{- if .Values.elasticsearch.persistence.enabled }}
```
Comment (frascuchon) [inline examples/deployments/k8s/argilla-chart/templates/NOTES.txt]: This should be also conditioned to the useOperator flag, right?

File: examples/deployments/k8s/argilla-chart/templates/NOTES.txt [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+Persistence is disabled. Your data will not be preserved across pod restarts.
+{{- end }}
+
+{{- if .Values.redis.enabled }}
+Redis is running as a dependency of this chart. You can access it at:
+  {{ .Release.Name }}-redis-master:6379
+{{- else }}
+Using external Redis at:
+  {{ .Values.externalRedis.host }}:{{ .Values.externalRedis.port }}
+{{- end }}
```
Comment (frascuchon) [inline examples/deployments/k8s/argilla-chart/templates/NOTES.txt]: Swapping redis notifications: first installation, then persistence storage.

```suggestion
{{- if .Values.redis.enabled }}
Redis is running as a dependency of this chart. You can access it at:
  {{ .Release.Name }}-redis-master:6379
{{- else }}
Using external Redis at:
  {{ .Values.externalRedis.host }}:{{ .Values.externalRedis.port }}
{{- end }}

{{- if .Values.redis.master.persistence.enabled }}
Persistence for Redis is enabled. Your Redis data will be stored in a Persistent Volume Claim named:
  {{ .Release.Name }}-redis-pvc
{{- end }}
{{- if and (not .Values.redis.master.persistence.enabled) (not .Values.elasticsearch.persistence.enabled) (not .Values.argilla.persistence.enabled)}}
Persistence is disabled. Your data will not be preserved across pod restarts.
{{- end }}
```

File: examples/deployments/k8s/argilla-chart/templates/NOTES.txt [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+Persistence for Argilla is enabled. Your Argilla data will be stored in a Persistent Volume Claim named:
+  {{ .Release.Name }}-argilla-pvc
+{{- end }}
+
+{{- if .Values.elasticsearch.persistence.enabled }}
+Persistence for Elasticsearch is enabled. Your Elasticsearch data will be stored in a Persistent Volume Claim named:
+  {{ .Release.Name }}-elasticsearch-pvc
+{{- end }}
+
+{{- if .Values.redis.master.persistence.enabled }}
```
Comment (frascuchon) [inline examples/deployments/k8s/argilla-chart/templates/NOTES.txt]: This should be conditioned also to redis.enabled flag, right?

File: examples/deployments/k8s/argilla-chart/values.yaml [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    accessMode: ReadWriteOnce
+    size: 2Gi
+    storageClass: "standard"
+    mountPath: "/data"
+  ingress:
+    enabled: true
+    className: "nginx"
+    host: "argilla.local"
+  configmap:
+    minikubeIP: "192.168.64.2"
```
Comment (frascuchon) [inline examples/deployments/k8s/argilla-chart/values.yaml]: Are these a values example for a specific environment?

**Review by frascuchon — COMMENTED** [review]
Thanks @bikash119  for this amazing PR.

I let some comments/questions. I will try the installation later.

File: examples/deployments/k8s/argilla-chart/templates/NOTES.txt (hunk 6 above)
Comment (bikash119) [reply examples/deployments/k8s/argilla-chart/templates/NOTES.txt]: Thank you @frascuchon for reviewing my work. I have incorporated these suggestions.

File: examples/deployments/k8s/argilla-chart/templates/NOTES.txt (hunk 8 above)
Comment (bikash119) [reply examples/deployments/k8s/argilla-chart/templates/NOTES.txt]: Thank you @frascuchon for reviewing my work. I have incorporated these suggestions.

File: examples/deployments/k8s/argilla-chart/values.yaml (hunk 9 above)
Comment (bikash119) [reply examples/deployments/k8s/argilla-chart/values.yaml]: To test locally, I used Minikube as the Kubernetes cluster. To simulate an external Elasticsearch (ES), I installed ES and accessed it via ingress. I used a ConfigMap to add entries to `/etc/hosts`, allowing Argilla to resolve the external ES host.
ConfigMap resource yaml 
```
{{- if not .Values.elasticsearch.useOperator }}
apiVersion: v1
kind: ConfigMap
metadata:
  name: custom-hosts
data:

  hosts: |
    127.0.0.1 localhost
    ::1 localhost
    {{ .Values.argilla.configmap.minikubeIP }} {{ .Values.argilla.ingress.host }}
{{- end }}
```

**Comment by jfcalvo** [comment]
@bikash119 additionally to the addition of Redis as dependency we need a change to run the Argilla server workers to process the background jobs enqueues into Redis.

You can take a look to https://github.com/argilla-io/argilla/blob/develop/examples/deployments/docker/docker-compose.yaml#L36-L47 to check how we are running the workers in the Argilla Docker compose example, but the idea is that the instance running the workers should have access to the same config that the server is using and then execute the following:

```sh
python -m argilla_server worker
```

Thanks for the PR btw ❤️

**Comment by bikash119** [comment]
@jfcalvo Thank you for the pointer. I have included it as a different resource to the helm chart. 
Additionally, I have updated the README.md to test the deployment in a clean minikube cluster. Was able to execute the integration test using `pytest tests/integration`
cc: @davidberenstein1957 , @frascuchon

**Review by frascuchon — APPROVED** [review]
Thanks for the contribution @bikash119!!!

**Pull request merged**